### PR TITLE
Reorganize the core API

### PR DESCRIPTION
**Pull request by djmitche**
This is a refactoring so that the core API methods
 - Are methods on a `Server` type that owns the storage backend
 - Fetch the `Client` as necessary
 - Report a missing client in such a way that the server can decide whether or not to create the client and try again

Generally, I think this makes the `core` API more usable for other implementations.

### CODE REVIEW

File: core/src/server.rs [hunk 1]
```diff
@@ -133,172 +91,321 @@ impl SnapshotUrgency {
...
+        parent_version_id: VersionId,
+        history_segment: HistorySegment,
+    ) -> Result<(AddVersionResult, SnapshotUrgency), ServerError> {
+        let mut txn = self.storage.txn()?;
+        let client = txn
+            .get_client(client_id)?
+            .ok_or(ServerError::NoSuchClient)?;
+
+        log::debug!(
+            "add_version(client_id: {}, parent_version_id: {})",
```
Comment (Necior) [inline core/src/server.rs:159]: Depending on your taste, you might prefer `log::debug!("add_version(client_id: {client_id}, parent_version_id: {parent_version_id})");` instead.

**Review by Necior — APPROVED** [review]

File: core/src/server.rs [hunk 2]
```diff
@@ -133,172 +91,321 @@ impl SnapshotUrgency {
...
+            txn.new_client(client_id, Uuid::nil())?;
+            for vnum in 0..num_versions {
+                let parent_version_id = version_id;
+                version_id = Uuid::new_v4();
+                versions.push(version_id);
+                txn.add_version(
+                    client_id,
+                    version_id,
+                    parent_version_id,
+                    vec![0, 0, vnum as u8],
```
[the remark this answers is not in the document]
Comment (djmitche) [reply core/src/server.rs:357]: It's just a quick way to make some distinct data for each version, of which there are many less than 256. This is just a test utility. I'll add a comment.

File: core/src/server.rs [hunk 3]
```diff
@@ -133,172 +91,321 @@ impl SnapshotUrgency {
...
+        })?;
+        Ok((server, client_id, versions))
+    }
+
+    /// Utility function to check the results of an add_version call
+    fn av_success_check(
+        server: &Server,
+        client_id: Uuid,
+        existing_versions: &[Uuid],
+        result: (AddVersionResult, SnapshotUrgency),
```
[the remark this answers is not in the document]
Comment (djmitche) [reply core/src/server.rs:382]: Good idea, thanks!

**Comment by djmitche** [comment]
Thanks for the review! I'll get this merged now. Perhaps if I @ you for a few additional reviews you'll gain more familiarity?

**Pull request merged**